network/p2p/router_test.go

```diff
@@ -0,0 +1,425 @@
+						return response, nil
+					})
+
+				callback := func(actualChainID ids.ID, actualResponse []byte, err error) {
+					defer wg.Done()
+					require.NoError(t, err)
+					require.Equal(t, chainID, actualChainID)
+					require.Equal(t, response, actualResponse)
+				}
+
+				require.NoError(t, client.CrossChainAppRequest(context.Background(), chainID, request, callback))
+			},
+		},
+		{
+			name: "cross-chain app request failed",
+			requestFunc: func(t *testing.T, router *Router, client *Client, sender *common.MockSender, handler *mocks.MockHandler, wg *sync.WaitGroup) {
+				sender.EXPECT().SendCrossChainAppRequest(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Do(func(ctx context.Context, chainID ids.ID, requestID uint32, request []byte) {
+						go func() {
+							require.NoError(t, router.CrossChainAppRequestFailed(ctx, chainID, requestID))
+						}()
+					})
+
+				callback := func(actualChainID ids.ID, actualResponse []byte, err error) {
+					defer wg.Done()
+
+					require.ErrorIs(t, err, ErrAppRequestFailed)
+					require.Equal(t, chainID, actualChainID)
+					require.Nil(t, actualResponse)
+				}
+
+				require.NoError(t, client.CrossChainAppRequest(context.Background(), chainID, request, callback))
+			},
+		},
+		{
+			name: "app gossip",
+			requestFunc: func(t *testing.T, router *Router, client *Client, sender *common.MockSender, handler *mocks.MockHandler, wg *sync.WaitGroup) {
+				sender.EXPECT().SendAppGossip(gomock.Any(), gomock.Any()).
+					Do(func(ctx context.Context, gossip []byte) {
+						go func() {
+							require.NoError(t, router.AppGossip(ctx, nodeID, gossip))
+						}()
+					}).AnyTimes()
+				handler.EXPECT().
+					AppGossip(context.Background(), nodeID, request).
+					DoAndReturn(func(context.Context, ids.NodeID, []byte) error {
+						defer wg.Done()
+						return nil
+					})
+
+				require.NoError(t, client.AppGossip(context.Background(), request))
+			},
+		},
+		{
+			name: "app gossip specific",
+			requestFunc: func(t *testing.T, router *Router, client *Client, sender *common.MockSender, handler *mocks.MockHandler, wg *sync.WaitGroup) {
+				sender.EXPECT().SendAppGossipSpecific(gomock.Any(), gomock.Any(), gomock.Any()).
+					Do(func(ctx context.Context, nodeIDs set.Set[ids.NodeID], gossip []byte) {
+						for n := range nodeIDs {
+							nodeID := n
+							go func() {
+								require.NoError(t, router.AppGossip(ctx, nodeID, gossip))
+							}()
+						}
+					}).AnyTimes()
+				handler.EXPECT().
+					AppGossip(context.Background(), nodeID, request).
+					DoAndReturn(func(context.Context, ids.NodeID, []byte) error {
+						defer wg.Done()
+						return nil
+					})
+
+				require.NoError(t, client.AppGossipSpecific(context.Background(), set.Set[ids.NodeID]{nodeID: struct{}{}}, request))
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+			ctrl := gomock.NewController(t)
```

Comment:
there should be a line such as `defer ctrl.Finish()` otherwise the mock library will not properly enforce all expected constraints.